src/qilib/configuration_helper/adapters/virtual_awg_instrument_adapter.py

```diff
@@ -0,0 +1,112 @@
+"""Quantum Inspire library
+
+Copyright 2019 QuTech Delft
+
+qilib is available under the [MIT open-source license](https://opensource.org/licenses/MIT):
+
+Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+persons to whom the Software is furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+"""
+
+try:
+    from qtt.instrument_drivers.virtualAwg.virtual_awg import VirtualAwg
+except ImportError as e:
+    raise ImportError(
+        "Quantum Technology Toolbox, qtt, not installed. Please do 'pip install qtt' or install from source.") from e
+
+from typing import Dict, Optional
+
+from qilib.configuration_helper import InstrumentAdapter, InstrumentAdapterFactory
+from qilib.configuration_helper.adapters import SettingsInstrumentAdapter
+from qilib.utils import PythonJsonStructure
+
+from qilib.configuration_helper.adapters.constants import INSTRUMENTS, ADAPTER_CLASS_NAME, ADDRESS, CONFIG, SETTINGS, \
+    AWG_MAP
+
+
+class VirtualAwgInstrumentAdapter(InstrumentAdapter):
+    def __init__(self, address: str) -> None:
+        super().__init__(address)
+
+        self._instrument: VirtualAwg = VirtualAwg()
+        self._adapters: Dict[str, InstrumentAdapter] = {}
+        self._settings_adapter: Optional[SettingsInstrumentAdapter] = None
+
+    def apply(self, config: PythonJsonStructure) -> None:
+        instruments = config[INSTRUMENTS]
+        for instrument in instruments:
+            adapter_class_name = instruments[instrument][ADAPTER_CLASS_NAME]
+            address = instruments[instrument][ADDRESS]
+            adapter_config = instruments[instrument][CONFIG]
+            self.add_instrument(adapter_class_name, address)
```

Comment:
This suffers from the same problem as the virtual dac: we only add instruments, so repeated calls to apply, but with different instruments in the config, will lead to the union of instruments in all configs to be present in the virtual awg, where one would expect that only the instruments specified in the config will be present.

For example, let's say we apply first a config that adds AWG1 and then AWG2. Fine, virtual awg has `[AWG1, AWG2]`. Now we add a config that has AWG2 and AWG3. I would expect the virtual awg to have `[AWG2, AWG3]` now, yet what it will actually contain is `[AWG1, AWG2, AWG3]`. That's not what should happen, should it?